cmd/istiod/istiod.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"istio.io/istio/pkg/istiod"
+	"istio.io/istio/pkg/istiod/k8s"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"log"
+	"os"
+)
+
+// Istio control plane with K8S support.
+//
+// - config is loaded from local K8S (in-cluster or using KUBECONFIG)
+// - local endpoints and additional registries from K8S
+// - additional MCP registries supported
+// - includes a Secret controller that provisions certs as secrets.
+//
+// Normal hyperistio is using local config files and MCP sources for config/endpoints,
+// as well as SDS backed by a file-based CA.
+func main() {
+	stop := make(chan struct{})
+
+	// First create the k8s clientset - and return the config source.
+	// The config includes the address of apiserver and the public key - which will be used
+	// after cert generation, to check that Apiserver-generated certs have same key.
+	client, kcfg, err := k8s.CreateClientset(os.Getenv("KUBECONFIG"), "")
+	if err != nil {
+		log.Fatal("Failed to connect to k8s", err)
+	}
+
+	// Load the mesh config. Note that the path is slightly changed - attempting to move all istio
+	// related under /var/lib/istio, which is also the home dir of the istio user.
+	istiods, err := istiod.InitConfig("/var/lib/istio/config")
+	if err != nil {
+		log.Fatal("Failed to start istiod ", err)
+	}
+
+	// Create k8s-signed certificates. This allows injector, validation to work without Citadel, and
+	// allows secure SDS connections to Istiod.
+	initCerts(istiods, client, kcfg)
+
+	// Init k8s related components, including Galley K8S controllers and
+	// Pilot discovery. Code kept in separate package.
+	k8sServer, err := k8s.InitK8S(istiods, client, kcfg, istiods.Args)
+	if err != nil {
+		log.Fatal("Failed to start k8s controllers ", err)
+	}
+
+	// Initialize Galley config source for K8S.
+	galleyK8S, err := k8sServer.NewGalleyK8SSource(istiods.Galley.Resources)
+	istiods.Galley.Sources = append(istiods.Galley.Sources, galleyK8S)
+
+	err = istiods.InitDiscovery()
+	if err != nil {
+		log.Fatal("Failed to init XDS server ", err)
```

Comment:
at some point need to use istio logging for consistency

Reply:
I'll fix it, we use istio logging in most places, I think only Fatal is not.